cleanlab/rank.py

```diff
@@ -541,3 +542,49 @@ def get_confidence_weighted_entropy_for_each_label(
     label_quality_scores = np.log(label_quality_scores + 1) / clipped_scores
 
     return label_quality_scores
+
+
+def get_knn_distance_ood_scores(
+    query_features: np.array, nbrs: NearestNeighbors, k: int
+) -> np.array:
+    """Returns the KNN distance out-of-distribution (OOD) score for each datapoint.
+
+    This is a function to compute OOD scores where higher scores indicate the datapoint is more likely to be OOD.
+
+    Parameters
+    ----------
+    query_features : np.array
+      Feature matrix of shape ``(N, ...)``, where N is the number of datapoints.
+      This is the "query set" of features for each datapoint which are used for nearest neighbor search.
+
+    nbrs : sklearn.neighbors.NearestNeighbors
```

Comment:
Is it possible to allow `nbrs` to be None?  (not necessarily by default)

If it is None, then we could construct the `nbrs` from the passed in `features` via sklearn.nearestneighbors inside of this function.

In particular if I just want to find OOD points in my training data, it appears I currently have to compute the neighbors twice, once when constructing `nbrs` and once again inside this function when I pass in the same data as `features`.  It would be nice if `nbrs` is None, then we only need to compute neighbors once for the `features` to construct `nbrs` and simultaneously compute their OOD score.

Reply:
Yeah it's possible and makes sense if we want to construct `nbrs` using the `features` we pass into `get_knn_distance_ood_scores`.

In this case where we want to find OOD points in the training data: when computing the neighbor distances, we'll need to ignore the first neighbor since the nearest neighbor of each training point here is the point itself, with a distance of zero. We can perhaps add a `ignore_first_neighbor` boolean parameter if we want to ignore first neighbor when averaging distances.